### PR TITLE
Bug Fixes and exposing Array internals

### LINKED ISSUE #1131: Expose functions to provide information about the memory layout of arrays
- [x] Strides
- [x] isContinguous (isLinear?)
- [x] isOwner
- [x] deviceOffset (for OpenCL backend)
- [x] rawDevicePointer

### LINKED ISSUE #1177: Allow specifying the stride when creating arrays from device memory
It would be useful if one could pass stride argument to `af_device_array()`.

### LINKED ISSUE #1275: Bug in index + moddims
Reproducing code:

``` cpp
    float ha[] = {1,
                  101, 102, 103, 104, 105,
                  201, 202, 203, 204, 205,
                  301, 302, 302, 304, 305,
                  401, 402, 403, 404, 405,

                  1010, 1020, 1030, 1040, 1050,
                  2010, 2020, 2030, 2040, 2050,
                  3010, 3020, 3030, 3040, 3050,
                  4010, 4020, 4030, 4040, 4050};
    {   
        af::array aa(sizeof(ha) / sizeof(float), ha);
        af::array aa0 = aa(af::seq(1, af::end));
        af::array aa1 = af::moddims(aa0, 5, 4, 2);
        af::array aa2 = aa1(af::seq(3), af::seq(3), af::span);
        af_print(aa2);
    }
```

output:
```
aa2
[3 3 2 1]
  101.0000     0.0000    -0.1090
  102.0000     0.0000     0.0000
  103.0000     0.0150    79.2937

    0.0000    -0.1090 -3630856635678720.0000
    0.0000     0.0000     0.0000
    0.0150    79.2937     0.0000
```

Expected:
```
aa2
[3 3 2 1]
  101.0000     201.0000    301.0000
  102.0000     202.0000    302.0000
  103.0000     203.0000    303.0000

  1010.0000     2010.0000    3010.0000
  1020.0000     2020.0000    3020.0000
  1030.0000     2030.0000    3030.0000
```

### PR DESCRIPTION
**Pull request by pavanky**
- af_create_array_with_strides
- af_get_strides
- af_get_offset
- af_get_raw_ptr
- af_is_linear
- af_is_owner

------------------------
- Fixes #1131 
- Fixes #1177 
- Fixes #1275

### CODE REVIEW

**Comment by 9prady9** [comment]
@pavanky I see that new array construction is not included into class `array` in c++ level. Is it because of ABI breakage ? So that we can deprecate these later when we push them into the `class` for 4.0 ?

**Comment by FilipeMaia** [comment]
@pavanky Yay! :+1:

**Comment by pavanky** [comment]
@9prady9 I am using a function instead of a constructor.

File: src/api/unified/internal.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+af_err af_create_array_with_strides(af_array *arr,
+                                    const void *data,
+                                    const dim_t offset,
+                                    const unsigned ndims,
+                                    const dim_t *const dims_,
+                                    const dim_t *const strides_,
+                                    const af_dtype ty,
+                                    const af_source location)
+{
+    CHECK_ARRAYS(arr);
```
Comment (shehzan10) [inline src/api/unified/internal.cpp]: No check arrays here. It's output.

File: src/api/c/internal.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                                    const void *data,
+                                    const dim_t offset,
+                                    const unsigned ndims,
+                                    const dim_t *const dims_,
+                                    const dim_t *const strides_,
+                                    const af_dtype ty,
+                                    const af_source location)
+{
+    try {
+        ARG_ASSERT(2, offset >= 0);
```
Comment (shehzan10) [inline src/api/c/internal.cpp]: You'll need to add af_init here as it a creation function. Like https://github.com/arrayfire/arrayfire/blob/devel/src/api/c/data.cpp#L78

Comment (pavanky) [inline src/api/c/internal.cpp]: good catches :+1:

**Comment by pavanky** [comment]
@shehzan10 fixed and forced pushed. I'll be adding tests and docs soon. You can test after this.

**Comment by pavanky** [comment]
@FilipeMaia btw we are limiting the `strides[0] == 1`.

**Comment by FilipeMaia** [comment]
@pavanky Thanks, I assumed as much.

**Comment by pavanky** [comment]
build arrayfire ci

**Pull request merged**